lean/exercises/metam.lean

```diff
@@ -0,0 +1,422 @@
+b) `2 + 1 =?= 1 + 2`  
+c) `?a =?= 2`, where `?a` has a type `String`  
+d) `?a + Int =?= "hi" + ?b`, where `?a` and `?b` don't have a type  
+e) `2 + ?a =?= 3`  
+f) `2 + ?a =?= 2 + 1`
+-/
+
+-- a) `5 =?= (λ x => 5) ((λ y : Nat → Nat => y) (λ z : Nat => z))`
+-- Definitionally equal.
+def expr2 := (λ x => 5) ((λ y : Nat → Nat => y) (λ z : Nat => z))
+#eval show MetaM Unit from do
+  let expr1 := Lean.mkNatLit 5
+  let expr2 := Expr.const `expr2 []
+  let isEqual ← Lean.Meta.isDefEq expr1 expr2
+  IO.println isEqual -- true
+
+-- b) `2 + 1 =?= 1 + 2`
+-- Definitionally equal.
+#eval show MetaM Unit from do
+  let expr1 := Lean.mkAppN (Expr.const `Nat.add []) #[Lean.mkNatLit 2, Lean.mkNatLit 1]
+  let expr2 := Lean.mkAppN (Expr.const `Nat.add []) #[Lean.mkNatLit 1, Lean.mkNatLit 2]
+  let isEqual ← Lean.Meta.isDefEq expr1 expr2
+  IO.println isEqual -- true
+
+-- c) `?a =?= 2`, where `?a` has a type `String`
+-- Not definitionally equal.
+#eval show MetaM Unit from do
+  let expr1 ← Lean.Meta.mkFreshExprMVar (Expr.const `String []) (userName := `expr1)
+  let expr2 := Lean.mkNatLit 2
+  let isEqual ← Lean.Meta.isDefEq expr1 expr2
+  IO.println isEqual -- false
+
+-- d) `?a + Int =?= "hi" + ?b`, where `?a` and `?b` don't have a type
+-- Definitionally equal.
+-- `?a` is assigned to `"hi"`, `?b` is assigned to `Int`.
+#eval show MetaM Unit from do
+  let a ← Lean.Meta.mkFreshExprMVar Option.none (userName := `a)
+  let b ← Lean.Meta.mkFreshExprMVar Option.none (userName := `b)
+  let expr1 := Lean.mkAppN (Expr.const `Nat.add []) #[a, Expr.const `Int []]
+  let expr2 := Lean.mkAppN (Expr.const `Nat.add []) #[Lean.mkStrLit "hi", b]
+  let isEqual ← Lean.Meta.isDefEq expr1 expr2
+  IO.println isEqual -- true
+
+  IO.println s!"a: {← instantiateMVars a}"
+  IO.println s!"b: {← instantiateMVars b}"
+
+-- e) `2 + ?a =?= 3`
+-- Not definitionally equal.
+#eval show MetaM Unit from do
+  let a ← Lean.Meta.mkFreshExprMVar (Expr.const `Nat []) (userName := `a)
+  let expr1 := Lean.mkAppN (Expr.const `Nat.add []) #[Lean.mkNatLit 2, a]
+  let expr2 := Lean.mkNatLit 3
+  let isEqual ← Lean.Meta.isDefEq expr1 expr2
+  IO.println isEqual -- false
+
+-- f) `2 + ?a =?= 2 + 1`
+-- Definitionally equal.
+-- `?a` is assigned to `1`.
+#eval show MetaM Unit from do
+  let a ← Lean.Meta.mkFreshExprMVar (Expr.const `Nat []) (userName := `a)
+  let expr1 := Lean.mkAppN (Expr.const `Nat.add []) #[Lean.mkNatLit 2, a]
+  let expr2 := Lean.mkAppN (Expr.const `Nat.add []) #[Lean.mkNatLit 2, Lean.mkNatLit 1]
+  let isEqual ← Lean.Meta.isDefEq expr1 expr2
+  IO.println isEqual -- true
+
+  IO.println s!"a: {← instantiateMVars a}"
+
+/- 9. [**Computation**] Write down what you expect the following code to output.
+
+```
+@[reducible] def reducibleDef     : Nat := 1
+@[instance] def instanceDef       : Nat := 2 -- same as `instance`
+def defaultDef                    : Nat := 3
+@[irreducible] def irreducibleDef : Nat := 4 -- same as `abbrev`
```

Comment:
Omg misplaced, thanks!

Reply:
> whether `abbrev` is its own special thing or an alias for `@[reducible]`

Probably ok to leave it like that in exercises even if we're uncertain, because we say "Note that `abbrev` is a shorthand for `@[reducible] def`" in the main text.